File: .ipynb_checkpoints/linear_model-checkpoint.r

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 3) {
    stop("Usage: Rscript linear_model.r <filename> <x_col> <y_col>")}

filename <- args[1]
x_col <- args[2]
y_col <- args[3]

dataset <- read.csv(filename)

formula <- as.formula(paste(y_col, "~", x_col))
model <- lm(formula, data=dataset)
model_summary <- summary(model)
slope <- coef(model)[2]
intercept <- coef(model)[1]
r2 <- model_summary$r.squared
r <- cor(dataset[[x_col]], dataset[[y_col]])
mse <- mean((predict(model) - dataset[[y_col]])^2)

dataset$predicted <- predict(model, newdata = dataset)

library(ggplot2)

dataset$Group <- "Data points"

plot <- ggplot(dataset, aes(x = .data[[x_col]], y = .data[[y_col]])) +
geom_point(aes(color = Group)) +
geom_line(aes(y = predicted, color = "Fitted line")) +
annotate("text", x = 2, y = 60000, col = "navy",
         label = paste(
            "y =", round(slope, 2),"x +", round(intercept,2), "\n",
            "r2:", round(r2, 2), "\n",
            "r:", round(r, 2), "\n",
            "MSE:", round(mse, 2), "\n")) +             
ggtitle(paste(y_col, "vs", x_col)) +
xlab(x_col) +
ylab(y_col) +
scale_color_manual(
    name = "Legend",
    values = c("Data points" = "green", 
               "Fitted line" = "purple")) 

ggsave("regression_plot_r.png", plot)
print(plot)

cat(
"Slope:", slope, "\n",
"Intercept:", intercept, "\n",
"r2:", r2, "\n",
"r:", r, "\n",
"MSE:", mse, "\n")

File: .ipynb_checkpoints/linear_model-checkpoint.py
#!/usr/bin/env python
# coding: utf-8

# # Linear Regression
# This script demonstrates a linear regression analysis using Python to model Salary based on Years of Experience.

import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error


if len(sys.argv) != 4:
    print("Usage: python linear_model.py <filename> <x_col> <y_col>")
    sys.exit(1)

filename = sys.argv[1]
x_col = sys.argv[2]
y_col = sys.argv[3]


dataset = pd.read_csv(filename)


X = dataset[[x_col]]
Y = dataset[[y_col]]


model = LinearRegression()
model.fit(X, Y)


model.score(X, Y)


slope = model.coef_[0]
slope = slope.item()

intercept = model.intercept_
intercept = intercept.item()


y_predicted = model.predict(X)


mse = mean_squared_error(Y,y_predicted)


r2 = model.score(X, Y)
r = dataset[x_col].corr(dataset[y_col])

plt.scatter(X, Y, color = "magenta", label = "Data points")
plt.plot(X, y_predicted, color = "teal", label = "Fitted line")
metrics_text = (
    f"y = {slope:.3f}x + {intercept:.3f}\n"
    f"r2 = {r2:.3f}\nr = {r:.3f}\nMSE = {mse:.3f}")
plt.text(1, 55000, metrics_text, fontsize = 10, color = "purple")
plt.title(f'{y_col} vs {x_col}')
plt.xlabel(x_col)
plt.ylabel(y_col)
plt.legend()
plt.savefig("regression_plot_python.png")
plt.show()


print(f"slope is {slope}")
print(f"intercept is {intercept}")
print(f"mse is {mse}")
print(f"r2 is {r2}")
print(f"r is {r}")

